pkg/applicationserver/metadata/location_registry.go

```diff
@@ -0,0 +1,245 @@
+		EndDeviceIdentifiers: ids,
+		FieldMask:            endDeviceLocationFieldMask,
+	}, c.WithClusterAuth())
+	if err != nil {
+		return nil, err
+	}
+	return dev.Locations, nil
+}
+
+// Merge implements EndDeviceLocationRegistry.
+func (c clusterEndDeviceLocationRegistry) Merge(ctx context.Context, ids ttnpb.EndDeviceIdentifiers, update map[string]*ttnpb.Location) (map[string]*ttnpb.Location, error) {
+	cc, err := c.GetPeerConn(ctx, ttnpb.ClusterRole_ENTITY_REGISTRY, nil)
+	if err != nil {
+		return nil, err
+	}
+	cl := ttnpb.NewEndDeviceRegistryClient(cc)
+	ctx, cancel := context.WithTimeout(ctx, c.timeout)
+	defer cancel()
+	dev, err := cl.Get(ctx, &ttnpb.GetEndDeviceRequest{
+		EndDeviceIdentifiers: ids,
+		FieldMask:            endDeviceLocationFieldMask,
+	}, c.WithClusterAuth())
+	if err != nil {
+		return nil, err
+	}
+	if len(update) == 0 {
+		return dev.Locations, nil
+	}
+	if dev.Locations == nil {
+		dev.Locations = make(map[string]*ttnpb.Location, len(update))
+	}
+	for k, l := range update {
+		dev.Locations[k] = l
+	}
+	_, err = cl.Update(ctx, &ttnpb.UpdateEndDeviceRequest{
+		EndDevice: ttnpb.EndDevice{
+			EndDeviceIdentifiers: ids,
+			Locations:            dev.Locations,
+		},
+		FieldMask: endDeviceLocationFieldMask,
+	}, c.WithClusterAuth())
+	if err != nil {
+		return nil, err
+	}
+	return dev.Locations, nil
+}
+
+// NewClusterEndDeviceLocationRegistry returns an EndDeviceLocationRegistry connected to the Entity Registry.
+func NewClusterEndDeviceLocationRegistry(cluster ClusterPeerAccess, timeout time.Duration) EndDeviceLocationRegistry {
+	return &clusterEndDeviceLocationRegistry{
+		ClusterPeerAccess: cluster,
+		timeout:           timeout,
+	}
+}
+
+type cachedEndDeviceLocationRegistry struct {
+	registry EndDeviceLocationRegistry
+	cache    EndDeviceLocationCache
+
+	softTTL time.Duration
+	hardTTL time.Duration
+
+	replicationPool workerpool.WorkerPool
+}
+
+// Get implements EndDeviceLocationRegistry.
+func (c *cachedEndDeviceLocationRegistry) Get(ctx context.Context, ids ttnpb.EndDeviceIdentifiers) (map[string]*ttnpb.Location, error) {
+	locations, storedAt, err := c.cache.Get(ctx, ids)
+	switch {
+	case err != nil && !errors.IsNotFound(err):
+		return nil, err
+	case err != nil && errors.IsNotFound(err):
+		locations = nil
+	case err == nil:
+		age := time.Since(*storedAt)
+		if age <= c.softTTL {
+			// If the object is younger than the soft TTL, just return the cached value.
+			return locations, nil
+		}
+		if remaining := c.hardTTL - age; remaining > 0 {
+			// If the objects age is between the soft and hard TTL, check if we should asynchronously
+			// refresh the cache.
+			window := c.hardTTL - c.softTTL
+			threshold := time.Duration(random.Int63n(int64(window)))
+			// remaining is the remaining window of the soft TTL in the (0, window) interval.
+			// threshold is a uniformly distributed duration in the [0, window) interval.
+			if remaining >= threshold {
+				return locations, nil
+			}
+		}
+	}
```

Comment:
I've updated the refreshing logic. Now instead of actually having expiring keys, we just guarantee that we will do a cache refresh after the hard TTL has passed. So now:

- Items younger than the soft TTL are just served as is.
- Items whose age is between the soft TTL and hard TTL may trigger an asynchronous cache refresh.
- Items whose age is higher than the hard TTL are guaranteed to trigger an asynchronous cache refresh.